### PR TITLE
[Jenkins-52906] add support for jcasc

### PR DESCRIPTION
**Pull request by casz**
[JENKINS-52906](https://issues.jenkins-ci.org/browse/JENKINS-52906)
https://github.com/jenkinsci/configuration-as-code-plugin/issues/809

* Convert global configuration to an actual Global Configuration extension
* Populate jira sites from folders in Job configuration.
* ~Make URL the only mandatory field and simplify constructor. See #173~ 
* Use List's instead of Array's, when possible
* ~Use numbers in jira site's jelly. See #172~
* Fix tests:
  * Add jcasc test.
  * Convert jira folder property test to use round trip test, caught a few mistakes in jira site setters.
  * ~faster travis build & Use enable Jacoco flag. See #171~

This is easier to shown in UI, we now support folder in folder lookup of Jira Site 🙌
image
Hmm wonder if it would be better to sort the list 😅

Nice jcasc doc:
image

image

Nice jcasc export:
```yaml
plugins:
  sites:
  - id: "default"
    url: "http://updates.jenkins-ci.org/update-center.json"
jenkins:
  agentProtocols:
  - "CLI-connect"
  - "CLI2-connect"
  - "JNLP-connect"
  - "JNLP2-connect"
  - "JNLP4-connect"
  - "Ping"
  disableRememberMe: false
  mode: NORMAL
  numExecutors: 2
  primaryView:
    all:
      name: "all"
  quietPeriod: 5
  scmCheckoutRetryCount: 0
  slaveAgentPort: 0
  views:
  - all:
      name: "all"
unclassified:
  jiraglobalconfiguration:
    sites:
    - url: "https://issues.jenkins-ci.org/"
    - url: "https://backwardsCompatURL.com/"
tool:
  jdk:
    defaultProperties:
    - installSource:
        installers:
        - jdkInstaller:
            acceptLicense: false
```

### CODE REVIEW

**Comment by casz** [comment]
Writing more tests to fix coverage 😅😤😇

**Comment by artkoshelev** [comment]
test

**Comment by artkoshelev** [comment]
test comment

**Comment by casz** [comment]
@artkoshelev github had a datastore issue: https://status.github.com
Could you cleanup your test comments 😅
I'll look into review

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java [hunk 1]
```diff
@@ -31,42 +27,30 @@
...
+        return sites;
     }
 
     /**
      * Adds a JIRA site.
      *
      * @param site the JIRA site
      */
-    @DataBoundSetter
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: I would consider removing this all together.

File: src/main/java/hudson/plugins/jira/JiraProjectProperty.java [hunk 2]
```diff
@@ -41,36 +48,44 @@ public JiraProjectProperty(String siteName) {
...
-        return null;
-    }
 
-    @Override
-    public DescriptorImpl getDescriptor() {
-        return DESCRIPTOR;
+        ItemGroup itemGroup = owner.getParent();
+        if (itemGroup instanceof AbstractFolder<?>) {
+            do {
+                AbstractFolder<?> folder = (AbstractFolder<?>) itemGroup;
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraProjectProperty.java]: i would suggest move that block to separate method to simplify code and testing

Comment (artkoshelev) [reply src/main/java/hudson/plugins/jira/JiraProjectProperty.java]: And again, it's the same code as below, so we should reuse it

File: src/main/java/hudson/plugins/jira/JiraSite.java [hunk 3]
```diff
@@ -388,6 +402,80 @@ public boolean isAppendChangeTimestamp() {
...
+    @DataBoundSetter
+    public void setRecordScmChanges(boolean recordScmChanges) {
+        this.recordScmChanges = recordScmChanges;
+    }
+
+    @DataBoundSetter
+    public void setUserPattern(String userPattern) {
+        this.userPattern = Util.fixEmptyAndTrim(userPattern);
+
+        if (this.userPattern != null) {
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraSite.java]: using positive conditions and 'fail fast' approach improves code readability, e.g.
```
if (this.userPattern == null) {
    return null;
}

return Pattern.compile(this.userPattern);
```

File: src/main/java/hudson/plugins/jira/JiraSite.java [hunk 4]
```diff
@@ -267,13 +269,21 @@ public JiraSite(URL url, URL alternativeUrl, StandardUsernamePasswordCredentials
...
     public JiraSite(URL url, URL alternativeUrl, String credentialsId, boolean supportsWikiStyleComment, boolean recordScmChanges, String userPattern,
                     boolean updateJiraIssueForAllStatus, String groupVisibility, String roleVisibility, boolean useHTTPAuth, int timeout, int readTimeout, int threadExecutorNumber){
         this(url, alternativeUrl, CredentialsHelper.lookupSystemCredentials(credentialsId, url), supportsWikiStyleComment, recordScmChanges, userPattern,
              updateJiraIssueForAllStatus, groupVisibility, roleVisibility, useHTTPAuth, timeout, readTimeout, threadExecutorNumber);
     }
 
+    @DataBoundConstructor
+    public JiraSite(String url){
+        URL mainURL = toURL(url);
+        if (mainURL == null) throw new AssertionError("URL cannot be null");
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraSite.java]: your `toUrl` method would throw `AssetionError` itself, so no need to null check here

File: src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import jenkins.model.Jenkins;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundSetter;
+
+@Extension
+@Symbol("jira")
+public class JiraGlobalConfiguration extends GlobalConfiguration {
+
+    @Nonnull
+    public static JiraGlobalConfiguration getInstance() {
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java]: any reason why it's static?

File: src/test/java/hudson/plugins/jira/ConfigAsCodeTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class ConfigAsCodeTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void should_support_configuration_as_code() throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/ConfigAsCodeTest.java]: why it's the only place you decided to use snake case? Oo

File: src/test/java/hudson/plugins/jira/JiraFolderPropertyTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.List;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class JiraFolderPropertyTest {
+    @Rule
+    public final JenkinsRule r = new JenkinsRule();
+
+    @Test public void configRoundtrip() throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraFolderPropertyTest.java]: annotation on a separate line

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        jiraFolderProperty = new JiraFolderProperty();
+        firstList = new ArrayList<>();
+        jiraSite = new JiraSite("https://first.com/");
+        firstList.add(jiraSite);
+        jiraFolderProperty.setSites(firstList);
+        folder.getProperties().add(jiraFolderProperty);
+    }
+
+    @Test
+    public void testGetSitesNullWithoutFolder() throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: you don't need to prefix tests with `test`

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+    @Rule
+    public final JenkinsRule r = new JenkinsRule();
+
+    FreeStyleProject freeStyleProject;
+    Folder folder;
+    JiraFolderProperty jiraFolderProperty;
+    List<JiraSite> firstList;
+    JiraSite jiraSite;
+    @Before
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: nit: empty line here

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+public class JiraProjectPropertyTest {
+
+    @Rule
+    public final JenkinsRule r = new JenkinsRule();
+
+    FreeStyleProject freeStyleProject;
+    Folder folder;
+    JiraFolderProperty jiraFolderProperty;
+    List<JiraSite> firstList;
+    JiraSite jiraSite;
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: probably should not be global to improve test readability

File: .mvn/extensions.xml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+<extensions xmlns="http://maven.apache.org/EXTENSIONS/1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/EXTENSIONS/1.0.0 http://maven.apache.org/xsd/core-extensions-1.0.0.xsd">
```
Comment (olamy) [inline .mvn/extensions.xml]: not related should be part of an other pr

File: .mvn/maven.config [hunk 12]
```diff
@@ -0,0 +1,2 @@
+-Pconsume-incrementals
```
Comment (olamy) [inline .mvn/maven.config]: not related should be part of an other pr

File: .travis.yml [hunk 13]
```diff
@@ -5,6 +5,15 @@ sudo: false
 jdk:
   - openjdk8
 
+install:
```
Comment (olamy) [inline .travis.yml]: not related should be part of an other pr

File: pom.xml [hunk 14]
```diff
@@ -4,18 +4,20 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>3.5</version>
+    <version>3.25</version>
     <relativePath />
   </parent>
 
   <artifactId>jira</artifactId>
-  <version>3.0.4-SNAPSHOT</version>
```
Comment (olamy) [inline pom.xml]: not related should be part of an other pr

File: pom.xml [hunk 15]
```diff
@@ -115,24 +117,16 @@
         <groupId>org.eluder.coveralls</groupId>
         <artifactId>coveralls-maven-plugin</artifactId>
         <version>4.3.0</version>
+        <configuration>
```
Comment (olamy) [inline pom.xml]: why?

File: pom.xml [hunk 16]
```diff
@@ -115,24 +117,16 @@
...
-        <version>0.8.1</version>
-        <executions>
-          <execution>
-            <id>prepare-agent</id>
-            <goals>
-              <goal>prepare-agent</goal>
-            </goals>
-            <configuration>
-              <excludes>
-                <exclude>hudson/plugins/jira/Messages*.class</exclude>
```
Comment (olamy) [inline pom.xml]: why removing this?

File: src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import hudson.util.PersistedList;
+import java.util.List;
+import javax.annotation.Nonnull;
+import jenkins.model.GlobalConfiguration;
+import jenkins.model.Jenkins;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundSetter;
+
+@Extension
+@Symbol("jira")
```
Comment (olamy) [inline src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java]: A bit too generic name IMHO.

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java [hunk 18]
```diff
@@ -31,42 +27,30 @@
...
-        sites.replaceBy(req.bindJSONToList(JiraSite.class, formData.get("sites")));
-        return this;
-    }
-
     /**
      * Return the JIRA sites.
      *
      * @return the JIRA sites
      */
-    public JiraSite[] getSites() {
```
Comment (olamy) [inline src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: NO do not break backward compat they might be plugins using that.

File: src/main/resources/hudson/plugins/jira/JiraFolderProperty/config.jelly [hunk 19]
```diff
@@ -5,4 +5,4 @@
       <f:repeatableProperty field="sites"/>
     </f:entry>
   </f:section>
-</j:jelly>
```
Comment (olamy) [inline src/main/resources/hudson/plugins/jira/JiraFolderProperty/config.jelly]: ??

File: src/main/resources/hudson/plugins/jira/JiraGlobalConfiguration/config.jelly [hunk 20]
```diff
@@ -5,4 +5,4 @@
       <f:repeatableProperty field="sites"/>
     </f:entry>
   </f:section>
-</j:jelly>
\ No newline at end of file
+</j:jelly>
```
Comment (olamy) [inline src/main/resources/hudson/plugins/jira/JiraGlobalConfiguration/config.jelly]: ??

File: src/main/resources/hudson/plugins/jira/JiraSite/config.jelly [hunk 21]
```diff
@@ -28,13 +28,13 @@
     <c:select />
   </f:entry>
   <f:entry title="${%Connection timeout}" field="timeout">
-    <f:textbox default="10" />
```
Comment (olamy) [inline src/main/resources/hudson/plugins/jira/JiraSite/config.jelly]: not related should be part of an other pr

File: .mvn/maven.config (hunk 12 above)
Comment (casz) [reply .mvn/maven.config]: Related I wanted the PR to be incremental so I could use it in jcasc

File: src/main/java/hudson/plugins/jira/JiraSite.java (hunk 4 above)
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraSite.java]: No Alternative URL is not mandatory.

File: src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java (hunk 17 above)
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java]: I can remove the symbol and then it would just be 
```yaml
unclassified:
  jiraglobalconfiguration:
    sites:
    - url: https://jira.com/
```

File: pom.xml (hunk 16 above)
Comment (casz) [reply pom.xml]: And that's why the travis file is related, I have `-P enable-jacoco` there. Generated files should not be part of the coverage report.

File: .mvn/extensions.xml (hunk 11 above)
Comment (casz) [reply .mvn/extensions.xml]: can you please provide with a `CONTRIBUTING.md` on how you want your PR's to be split up.

File: src/main/java/hudson/plugins/jira/JiraSite.java (hunk 4 above)
Comment (artkoshelev) [reply src/main/java/hudson/plugins/jira/JiraSite.java]: yep, sorry )

File: src/main/java/hudson/plugins/jira/JiraMailAddressResolver.java [hunk 22]
```diff
@@ -30,13 +30,14 @@ public String findMailAddressFor(User u) {
...
-        for (JiraSite site : JiraProjectProperty.DESCRIPTOR.getSites()) {
+        for (JiraSite site : JiraGlobalConfiguration.getInstance().getSites()) {
             JiraSession session = site.getSession();
             if (session == null) {
                 continue;
             }
 
-            com.atlassian.jira.rest.client.api.domain.User user = session.service.getUser(username);
+            com.atlassian.jira.rest.client.api.domain.User user = session.service
+                .getUser(username);
```
Comment (warden) [inline src/main/java/hudson/plugins/jira/JiraMailAddressResolver.java]: Let's not do partial reformatting as part of a functional PR. In other words, keep the current formatting for unrelated changes, if we want to change line lengths, let's do it for all the source code in a separate PR.

File: .mvn/extensions.xml (hunk 11 above)
Comment (warden) [reply .mvn/extensions.xml]: I think the general rule is one functional (independent) change = PR and it's for two reasons:
easier reviewing (geez, that PR is big!) and easier backward reference when tracking changes.
So here, definitely the first two bullet points seem to me as two separate independent PRs:
```
* Convert global configuration to an actual Global Configuration extension
* Populate jira sites from folders in Job configuration.
```
The rest looks like nice improvements to the coding style and code testability/readability and I guess they're fine to be together in one PR.

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java (hunk 18 above)
Comment (olamy) [reply src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: github is not the only place in the world where people write code :)

Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: They can always adapt when they update their version then 🙊

File: .mvn/extensions.xml (hunk 11 above)
Comment (casz) [reply .mvn/extensions.xml]: Global configuration improvement went in because that made it easier for me to introduce support and testing for Jcasc 😅

File: src/test/java/hudson/plugins/jira/ConfigAsCodeTest.java (hunk 6 above)
Comment (casz) [reply src/test/java/hudson/plugins/jira/ConfigAsCodeTest.java]: Copy paste 🤭🙈

File: .mvn/extensions.xml (hunk 11 above)
Comment (olamy) [reply .mvn/extensions.xml]: incremental build should be separate. 
well it's not a matter of having paperwork/procedure/etc... but as said @warden this make pr smaller to review and containing one change. So it's easier to revert if it breaks..

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java (hunk 18 above)
Comment (warden) [reply src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: I think we're discussing here basically which version this will be released as in terms of https://semver.org/. There are two sides here:
* one is the downstream backward compatibility (for plugins that inherit from this plugin).
* second is the upstream Jenkins compatibility (for the changes that were needed to adapt this plugin to Jenkins API changes)

For the open source contributions to work smoothly (where a downstream plugin might pick up the changes in half a year from the upstream), we should `@Deprecate` methods if possible to have a _transition period_ of adaptation.

File: src/main/resources/hudson/plugins/jira/JiraSite/config.jelly (hunk 21 above)
Comment (warden) [reply src/main/resources/hudson/plugins/jira/JiraSite/config.jelly]: Separate PR indeed, not related to JCASC compatibility.

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java (hunk 18 above)
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: and when there is no transition? You cannot have two method with different return types.

File: src/main/java/hudson/plugins/jira/JiraSite.java (hunk 3 above)
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraSite.java]: We are not returning on void

File: src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java (hunk 5 above)
Comment (casz) [reply src/main/java/hudson/plugins/jira/JiraGlobalConfiguration.java]: Cause a descriptor is usually in a static context, I cannot override get descriptor because I want to avoid calling `Jenkins.getInstance().getDescriptorOrDie(JiraGlobalConfiguration.class)` all over the place and I should not instantiate a descriptor, Jenkins does that job 😅

**Comment by casz** [comment]
This is where I can cry in a corner, rewriting tests to please reviewers 😅

**Comment by casz** [comment]
This is the compare window your looking for: https://github.com/casz/jira-plugin/compare/simpleConstructor...casz:JENKINS-52906

**Comment by ewelinawilkosz** [comment]
great job with this fix, where do we stand though? were the comments addressed?

**Comment by casz** [comment]
No, haven't had time and seems like the PRs are blocked by stats padding the coverage if the comments were to be addressed.

**Comment by kmcquade** [comment]
@casz @ewelinawilkosz @olamy @artkoshelev - any updates? We really need this. Thanks

**Comment by timja** [comment]
@kmcquade Parts of this PR have been extracted to others see the heading,
Currently the outstanding one is: https://github.com/jenkinsci/jira-plugin/pull/177
Once that's been merged this one probably needs a little cleanup, hopefully its not far off though

**Comment by casz** [comment]
if #177 can get merged I will rework this PR 😄

**Comment by casz** [comment]
Please take another look :smile:

File: src/main/java/hudson/plugins/jira/JiraFolderProperty.java [hunk 23]
```diff
@@ -58,15 +41,40 @@ public JiraFolderProperty() {
...
+                AbstractFolder<?> folder = (AbstractFolder<?>) itemGroup;
+                JiraFolderProperty jiraFolderProperty = folder.getProperties()
+                    .get(JiraFolderProperty.class);
+                if (jiraFolderProperty != null && jiraFolderProperty.getSites().length != 0) {
+                    result.addAll(Arrays.asList(jiraFolderProperty.getSites()));
+                }
+                itemGroup = folder.getParent();
+            }
+
+            if (!(itemGroup instanceof AbstractFolder<?>)) {
```
Comment (artkoshelev) [inline src/main/java/hudson/plugins/jira/JiraFolderProperty.java]: that should be inside `while` instead of null checking

File: src/test/java/hudson/plugins/jira/JiraGlobalConfigurationTest.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class JiraGlobalConfigurationTest {
+    @Rule
+    public final JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void migrateOldConfiguration() throws Exception {
+        String url = "https://backwardsCompatURL.com/";
+        String oldConfiguration = "<hudson.plugins.jira.JiraProjectProperty_-DescriptorImpl>\n"
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraGlobalConfigurationTest.java]: i'd prefer to read xml from xml file so you'd have proper IDE support for that

File: src/test/java/hudson/plugins/jira/JiraGlobalConfigurationTest.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+        DescriptorImpl instance = (DescriptorImpl)xstream.fromXML(oldConfiguration);
+        assertNotNull(instance);
+        assertNull(instance.sites);
+        JiraSite actual = JiraGlobalConfiguration.get().getSites().get(0);
+        assertEquals(url, actual.getName());
+        r.assertEqualDataBoundBeans(expected, actual);
+    }
+
+    @Test
+    public void testNullCase() throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraGlobalConfigurationTest.java]: not clear what's that test doing

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+        assertNotNull(property);
+        assertNull(property.getSite());
+    }
+
+    @Test
+    public void testNoInitialDefault() throws Exception {
+        JiraProjectProperty prop = new JiraProjectProperty(null);
+        setupGlobalConfig("single-site.yml");
+        JiraSite site = prop.getSite();
+        assertNotNull(site);
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: this check is redundant

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+        assertEquals("https://jira.com/", site.getName());
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: why site URL is equal to 'name'?

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    @Test
+    public void testDefault() throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: please avoid using word "test" in your test names (it doesn't add any useful information, we already know it's a test)
please give tests more descriptive names

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+        freeStyleProject = secondFolder.createProject(FreeStyleProject.class, "something");
+        // testing we can get value from folder above.
+        testProperty("single-site.yml", firstList.get(0), true);
+    }
+
+    private void testProperty(String input, JiraSite expected) throws Exception {
+        testProperty(input, expected, false);
+    }
+
+     private void testProperty(String input, JiraSite expected, boolean roundTrip) throws Exception {
```
Comment (artkoshelev) [inline src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: could you please get rid of that method?
1. the `roundTrip` property is really used only in one test, it would be better to move that part to that test
2. conditional logic here complicates reading and understanding of what is test doing
3. you have to jump back and forth through the code to understand each test

in overall it's better to copy-paste few lines and have stupidly straightforward test cases rather then making your tests as complex as your code

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java (hunk 27 above)
Comment (casz) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: since `getName` returns `url.toExternalForm()` and `getUrl` has the chance of returning alternativeUrl and I would either have to cast it to `toExternalForm` or cast the other equation to `new URL()`.

```
    public URL getUrl() {
        return Objects.firstNonNull(this.url, this.alternativeUrl);
    }
```

```
    public String getName() {
        return url.toExternalForm();
    }
```

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java (hunk 28 above)
Comment (casz) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: There I avoided the name test, I went ahead and removed it from all existing tests.
I will fix your other comments once I get time.

**Comment by casz** [comment]
I will fix the later — followed @artkoshelev review comment to remove test in front of the name.

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java (hunk 27 above)
Comment (artkoshelev) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: i mean it's confusing to call `getName` and receive URL when you expect to receive, well, name...

Comment (casz) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: I didn't write the method names. 😢

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java (hunk 26 above)
Comment (casz) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: my IDE disagrees with you :confused: 
image

**Comment by casz** [comment]
@olamy @artkoshelev I should have addressed all your concerns.

File: src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java (hunk 26 above)
Comment (casz) [reply src/test/java/hudson/plugins/jira/JiraProjectPropertyTest.java]: There suppressing our beloved IDE

**Comment by casz** [comment]
Conflict resolved

**Comment by casz** [comment]
@olamy @artkoshelev what are the chances of this getting merged?

**Comment by artkoshelev** [comment]
It looks good to me, but change is pretty big, did you performed any kind of manual/integration testing?

**Comment by casz** [comment]
I don't actively use Jira. I am just here fixing the JCasC support plus some more while I was at it.

perhaps @kmcquade or @timja can :sweat_smile:

**Comment by casz** [comment]
Removing all the test addition and import changes I am left with the significant change which are 152 additions and 99 deletions. So how big was this change?

**Comment by casz** [comment]
I did do some testing. Tested that configuration would be transferred to the new JiraGlobalConfiguration and tested that nested folder sites would show up. Did you not see the screenshots in the PR description?
Besides that everything else should continue to work.

**Comment by casz** [comment]
@olamy Would be good to get this in :)

**Comment by casz** [comment]
@olamy @artkoshelev users are asking for the export feature
